File: tests/test_service_worker.js
// tests/test_service_worker.js - Validation du Service Worker KLASORA
const fs = require('fs');
const path = require('path');
const assert = require('assert');

async function runTests() {
    console.log('=== TEST SERVICE WORKER KLASORA ===');

    const swPath = path.join(__dirname, '..', 'app', 'static', 'service-worker.js');
    const swCode = fs.readFileSync(swPath, 'utf-8');

    assert.ok(swCode.includes("const CACHE_NAME = 'klasora-cache-v14';"), 'CACHE_NAME doit etre v14');
    assert.ok(swCode.includes('const CACHE_VERSION = CACHE_NAME;'), 'CACHE_VERSION doit pointer sur CACHE_NAME');
    assert.ok(!swCode.includes('fetchWithTimeout'), 'fetchWithTimeout doit etre supprime');
    assert.ok(!swCode.includes('4000'), 'Le timeout de 4000ms doit etre supprime');
    console.log('OK 1 : version de cache et ancien timeout');

    assert.ok(swCode.includes("request.mode === 'navigate'"), 'Les navigations HTML doivent etre interceptees');
    assert.ok(swCode.includes('async function handleNavigation(request)'), 'handleNavigation doit exister');
    assert.ok(swCode.includes('offlineFallback()'), 'offlineFallback doit etre utilise');
    assert.ok(swCode.includes('return await fetch(request);'), 'Les pages HTML privees doivent etre network-only');
    assert.ok(!swCode.includes('await cache.put(request, networkResponse.clone());'), 'Les pages HTML privees ne doivent pas etre cachees');
    console.log('OK 2 : navigation network-only avec fallback offline');

    assert.ok(swCode.includes("url.pathname === '/logout'"), '/logout doit declencher une purge');
    assert.ok(swCode.includes("cacheName.startsWith('klasora-pages-')"), 'Les anciens caches de pages doivent etre purgeables');
    assert.ok(swCode.includes("url.pathname === '/api/connectivity'"), '/api/connectivity doit rester network-only');
    console.log('OK 3 : purge logout et connectivite reseau');

    assert.ok(swCode.includes("'/manifest.json'"), 'Le manifest public doit etre precache');
    assert.ok(swCode.includes("'/static/css/style.css?v=14'"), 'Le CSS principal v14 doit etre precache');
    assert.ok(swCode.includes("'/static/js/pwa.js?v=14'"), 'Le JS PWA v14 doit etre precache');
    assert.ok(swCode.includes("'/static/img/icons/icon-192x192.png'"), 'Icone iOS critique precachee');
    console.log('OK 4 : assets critiques precaches');

    assert.ok(swCode.includes('await new Promise(resolve => setTimeout(resolve, 800));'), 'Retry court anti-rebond Flask conserve');
    assert.ok(swCode.includes('new Response('), 'Fallback ultime sans page blanche doit exister');
    assert.ok(swCode.includes('Connexion Internet requise pour vous authentifier'), 'Message login offline clair attendu');
    console.log('OK 5 : fallback offline sans page blanche');

    console.log('\nTOUS LES TESTS DU SERVICE WORKER ONT REUSSI');
}

runTests().catch(err => {
    console.error('ECHEC DES TESTS :', err);
    process.exit(1);
});

File: tests/test_service_worker.py
# tests/test_service_worker.py - Validation PWA Service Worker KLASORA
import os
import subprocess

from app import create_app


class PwaRouteTestConfig:
    TESTING = True
    WTF_CSRF_ENABLED = False
    SQLALCHEMY_DATABASE_URI = "sqlite:///:memory:"
    SECRET_KEY = "test-service-worker"


def test_service_worker_static_analysis():
    """Verifie statiquement les regles critiques du Service Worker."""
    sw_path = os.path.join(os.path.dirname(__file__), '..', 'app', 'static', 'service-worker.js')
    assert os.path.exists(sw_path), "service-worker.js doit exister"

    with open(sw_path, 'r', encoding='utf-8') as f:
        content = f.read()

    assert "const CACHE_VERSION = 'klasora-static-v7';" in content
    assert "const PAGE_CACHE = 'klasora-pages-v7';" in content
    assert "const CACHE_VERSION = 'klasora-static-v4';" not in content

    assert "fetchWithTimeout" not in content
    assert "4000" not in content

    assert "async function handleNavigation(request)" in content
    assert "request.mode === 'navigate'" in content
    assert "await new Promise(resolve => setTimeout(resolve, 800));" in content
    assert "event.respondWith(handleNavigation(request));" in content
    assert "await cache.put(request, networkResponse.clone());" in content
    assert "offlineFallback()" in content
    assert "Connexion Internet requise pour vous authentifier" in content

    assert "BYPASS_NAVIGATION_CACHE" in content
    assert "'/login'" in content
    assert "'/logout'" in content
    assert "'/api/'" in content
    assert "CLEAR_USER_CACHE" in content
    assert "await caches.delete(PAGE_CACHE)" in content

    assert "allowedCaches" in content
    assert "caches.delete(cacheName)" in content


def test_offline_template_has_clear_login_message():
    template_path = os.path.join(os.path.dirname(__file__), '..', 'app', 'templates', 'offline.html')
    assert os.path.exists(template_path), "offline.html doit exister"

    with open(template_path, 'r', encoding='utf-8') as f:
        content = f.read()

    assert "Connexion Internet requise pour vous authentifier." in content
    assert "Vous etes hors connexion" in content
    assert "window.addEventListener('online', updateStatus)" in content


def test_pwa_public_routes_available():
    app = create_app(PwaRouteTestConfig)
    client = app.test_client()

    offline = client.get('/offline')
    assert offline.status_code == 200
    assert b'Connexion Internet requise pour vous authentifier.' in offline.data

    manifest = client.get('/manifest.json')
    assert manifest.status_code == 200
    assert manifest.mimetype in ('application/manifest+json', 'application/json')

    service_worker = client.get('/service-worker.js')
    assert service_worker.status_code == 200
    assert service_worker.headers.get('Service-Worker-Allowed') == '/'


def test_service_worker_runtime_node():
    """Execute les tests unitaires JS emulant le Service Worker sous Node.js."""
    test_js_path = os.path.join(os.path.dirname(__file__), 'test_service_worker.js')
    assert os.path.exists(test_js_path), "test_service_worker.js doit exister"

    result = subprocess.run(
        ['node', test_js_path],
        capture_output=True,
        text=True,
        encoding='utf-8'
    )
    assert result.returncode == 0, f"Les tests JS ont echoue:\nSTDOUT:\n{result.stdout}\nSTDERR:\n{result.stderr}"
    assert "TOUS LES TESTS DU SERVICE WORKER ONT REUSSI" in result.stdout
